File: Python/Programmers/playing_alone.py
def solution(cards):
    answer = []
    opened = [0 for _ in range(len(cards))]
    for i in range(len(cards)):
        if not opened[i]:
            cnt = 0
            next = cards[i] - 1
            while not opened[next]:
                cnt += 1
                opened[next] = 1
                next = cards[next] - 1
            answer.append(cnt)
    
    if len(answer) == 1:
        return 0
    else:
        answer.sort(reverse = True)

    return answer[0]*answer[1]
print(solution([8,6,3,7,2,5,1,4]))

File: Java/Programmers/playing_alone.java
package Programmers;

import java.util.ArrayList;
import java.util.List;
import java.util.Collections;

class PlayingAlone {
    public int solution(int[] cards) {
        List<Integer> answer = new ArrayList<Integer>();
        int[] opened = new int[cards.length];
        for (int i = 0; i < cards.length; i ++) {
            if (opened[i] != 1) {
                int cnt = 0;
                int next = cards[i] - 1;
                while (opened[next] != 1) {
                    cnt += 1;
                    opened[next] = 1;
                    next = cards[next] - 1;
                }
                answer.add(cnt);
             }
        }
        
        if (answer.size() == 1) {
            return 0;
        } else {
            Collections.sort(answer, Collections.reverseOrder());
        }

        return answer.get(0)*answer.get(1);
    }
}


